File: frontend/src/services/companyService.ts
import { api } from '@/utils/api';

export interface Company {
  id: string;
  company_name: string;
  company_code: string;
  email: string;
  phone?: string;
  website?: string;
  industry?: string;
  business_type?: string;
  tax_id?: string;
  registration_number?: string;
  address_line1?: string;
  address_line2?: string;
  city?: string;
  state?: string;
  postal_code?: string;
  country?: string;
  logo_url?: string;
  primary_color?: string;
  secondary_color?: string;
  default_currency: string;
  default_language: string;
  timezone: string;
  date_format: string;
  fiscal_year_start: string;
  tax_settings?: Record<string, any>;
  enabled_modules?: Record<string, boolean>;
  numbering_formats?: Record<string, string>;
  subscription_tier: string;
  status: string;
  trial_ends_at?: string;
  database_schema?: string;
  created_at: string;
  updated_at: string;
}

export interface CompanyRegistrationRequest {
  company_name: string;
  email: string;
  phone?: string;
  website?: string;
  industry?: string;
  business_type?: string;
  tax_id?: string;
  registration_number?: string;
  address_line1?: string;
  address_line2?: string;
  city?: string;
  state?: string;
  postal_code?: string;
  country?: string;
  default_currency?: string;
  default_language?: string;
  timezone?: string;
  fiscal_year_start?: string;
}

export interface CompanyUser {
  id: string;
  company_id: string;
  user_id: string;
  role: string;
  is_admin: boolean;
  is_active: boolean;
  permissions?: Record<string, any>;
  created_at: string;
}

export interface CompanySettings {
  id: string;
  company_id: string;
  chart_of_accounts_template?: string;
  approval_workflows?: Record<string, any>;
  integrations?: Record<string, any>;
  custom_fields?: Record<string, any>;
  notification_settings?: Record<string, any>;
  security_settings?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

/**
 * Company Service
 * Provides methods to interact with the company management API endpoints
 */
export default {
  /**
   * Register a new company
   * @param companyData - Company registration data
   * @returns Promise with the registered company
   */
  async registerCompany(companyData: CompanyRegistrationRequest) {
    return api.post('/company/register', companyData);
  },

  /**
   * Get company by ID
   * @param companyId - Company ID
   * @returns Promise with company details
   */
  async getCompany(companyId: string) {
    return api.get(`/company/${companyId}`);
  },

  /**
   * Update company information
   * @param companyId - Company ID
   * @param companyData - Updated company data
   * @returns Promise with updated company
   */
  async updateCompany(companyId: string, companyData: Partial<Company>) {
    return api.put(`/company/${companyId}`, companyData);
  },

  /**
   * List companies
   * @param status - Optional status filter
   * @param limit - Maximum number of records
   * @returns Promise with list of companies
   */
  async listCompanies(status?: string, limit: number = 100) {
    const params = new URLSearchParams();
    if (status) params.append('status', status);
    params.append('limit', limit.toString());

    return api.get(`/company/?${params.toString()}`);
  },

  /**
   * Add user to company
   * @param companyId - Company ID
   * @param userData - User data
   * @returns Promise with company user
   */
  async addUserToCompany(companyId: string, userData: {
    user_id: string;
    role?: string;
    is_admin?: boolean;
    permissions?: Record<string, any>;
  }) {
    return api.post(`/company/${companyId}/users`, userData);
  },

  /**
   * Get company users
   * @param companyId - Company ID
   * @param activeOnly - Show only active users
   * @returns Promise with list of company users
   */
  async getCompanyUsers(companyId: string, activeOnly: boolean = true) {
    return api.get(`/company/${companyId}/users?active_only=${activeOnly}`);
  },

  /**
   * Get user companies
   * @param userId - User ID
   * @param activeOnly - Show only active companies
   * @returns Promise with list of user companies
   */
  async getUserCompanies(userId: string, activeOnly: boolean = true) {
    return api.get(`/company/user/${userId}/companies?active_only=${activeOnly}`);
  },

  /**
   * Update company settings
   * @param companyId - Company ID
   * @param settings - Settings data
   * @returns Promise with updated settings
   */
  async updateCompanySettings(companyId: string, settings: Partial<CompanySettings>) {
    return api.put(`/company/${companyId}/settings`, settings);
  },

  /**
   * Get company settings
   * @param companyId - Company ID
   * @returns Promise with company settings
   */
  async getCompanySettings(companyId: string) {
    return api.get(`/company/${companyId}/settings`);
  },

  /**
   * Utility functions for company operations
   */
  utils: {
    /**
     * Format company status for display
     */
    formatStatus(status: string): string {
      const statusMap: Record<string, string> = {
        'active': 'Active',
        'inactive': 'Inactive',
        'suspended': 'Suspended',
        'trial': 'Trial'
      };
      
      return statusMap[status] || status.charAt(0).toUpperCase() + status.slice(1);
    },

    /**
     * Get status color for UI
     */
    getStatusColor(status: string): string {
      const colorMap: Record<string, string> = {
        'active': 'success',
        'inactive': 'warning',
        'suspended': 'error',
        'trial': 'info'
      };
      
      return colorMap[status] || 'grey';
    },

    /**
     * Format subscription tier
     */
    formatSubscriptionTier(tier: string): string {
      const tierMap: Record<string, string> = {
        'basic': 'Basic',
        'professional': 'Professional',
        'enterprise': 'Enterprise'
      };
      
      return tierMap[tier] || tier.charAt(0).toUpperCase() + tier.slice(1);
    },

    /**
     * Get available industries
     */
    getIndustries(): string[] {
      return [
        'Technology',
        'Healthcare',
        'Finance',
        'Manufacturing',
        'Retail',
        'Education',
        'Real Estate',
        'Construction',
        'Transportation',
        'Hospitality',
        'Professional Services',
        'Non-profit',
        'Government',
        'Other'
      ];
    },

    /**
     * Get available business types
     */
    getBusinessTypes(): string[] {
      return [
        'Corporation',
        'LLC',
        'Partnership',
        'Sole Proprietorship',
        'Non-profit',
        'Government',
        'Other'
      ];
    },

    /**
     * Get available currencies
     */
    getCurrencies(): Array<{ code: string; name: string; symbol: string }> {
      return [
        { code: 'USD', name: 'US Dollar', symbol: '$' },
        { code: 'EUR', name: 'Euro', symbol: '€' },
        { code: 'GBP', name: 'British Pound', symbol: '£' },
        { code: 'CAD', name: 'Canadian Dollar', symbol: 'C$' },
        { code: 'AUD', name: 'Australian Dollar', symbol: 'A$' },
        { code: 'JPY', name: 'Japanese Yen', symbol: '¥' },
        { code: 'CHF', name: 'Swiss Franc', symbol: 'CHF' },
        { code: 'CNY', name: 'Chinese Yuan', symbol: '¥' }
      ];
    },

    /**
     * Get available timezones
     */
    getTimezones(): Array<{ value: string; label: string }> {
      return [
        { value: 'UTC', label: 'UTC' },
        { value: 'America/New_York', label: 'Eastern Time (US)' },
        { value: 'America/Chicago', label: 'Central Time (US)' },
        { value: 'America/Denver', label: 'Mountain Time (US)' },
        { value: 'America/Los_Angeles', label: 'Pacific Time (US)' },
        { value: 'Europe/London', label: 'London' },
        { value: 'Europe/Paris', label: 'Paris' },
        { value: 'Europe/Berlin', label: 'Berlin' },
        { value: 'Asia/Tokyo', label: 'Tokyo' },
        { value: 'Asia/Shanghai', label: 'Shanghai' },
        { value: 'Australia/Sydney', label: 'Sydney' }
      ];
    },

    /**
     * Get available languages
     */
    getLanguages(): Array<{ code: string; name: string }> {
      return [
        { code: 'en-US', name: 'English (US)' },
        { code: 'en-GB', name: 'English (UK)' },
        { code: 'es-ES', name: 'Spanish' },
        { code: 'fr-FR', name: 'French' },
        { code: 'de-DE', name: 'German' },
        { code: 'it-IT', name: 'Italian' },
        { code: 'pt-BR', name: 'Portuguese (Brazil)' },
        { code: 'ja-JP', name: 'Japanese' },
        { code: 'zh-CN', name: 'Chinese (Simplified)' }
      ];
    },

    /**
     * Validate company code format
     */
    validateCompanyCode(code: string): { valid: boolean; message?: string } {
      if (!code) {
        return { valid: false, message: 'Company code is required' };
      }
      
      if (code.length < 3 || code.length > 10) {
        return { valid: false, message: 'Company code must be 3-10 characters' };
      }
      
      if (!/^[A-Z0-9]+$/.test(code)) {
        return { valid: false, message: 'Company code must contain only uppercase letters and numbers' };
      }
      
      return { valid: true };
    },

    /**
     * Check if company is in trial
     */
    isTrialCompany(company: Company): boolean {
      return company.status === 'trial';
    },

    /**
     * Get trial days remaining
     */
    getTrialDaysRemaining(company: Company): number {
      if (!company.trial_ends_at) return 0;
      
      const trialEnd = new Date(company.trial_ends_at);
      const now = new Date();
      const diffTime = trialEnd.getTime() - now.getTime();
      const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
      
      return Math.max(0, diffDays);
    },

    /**
     * Format address
     */
    formatAddress(company: Company): string {
      const parts = [
        company.address_line1,
        company.address_line2,
        company.city,
        company.state,
        company.postal_code,
        company.country
      ].filter(Boolean);
      
      return parts.join(', ');
    }
  }
};

File: backend/app/services/company/company_service.py
"""
Company management service.
"""
from datetime import datetime, timedelta
from typing import List, Optional, Dict, Any

from sqlalchemy import and_, desc
from sqlalchemy.orm import Session
from uuid import UUID
import secrets

from app.models.company import Company, CompanyUser, CompanySettings, CompanyStatus, SubscriptionTier





class CompanyService:
    """Service for managing company profiles and multi-tenant operations."""
    
    def __init__(self, db: Session):
        self.db = db
    
    def register_company(self, company_data: Dict[str, Any], admin_user_id: UUID) -> Company:
        company_code = self._generate_company_code(company_data['company_name'])
        
        company = Company(
            company_name=company_data['company_name'],
            company_code=company_code,
            email=company_data['email'],
            phone=company_data.get('phone'),
            website=company_data.get('website'),
            industry=company_data.get('industry'),
            business_type=company_data.get('business_type'),
            tax_id=company_data.get('tax_id'),
            registration_number=company_data.get('registration_number'),
            address_line1=company_data.get('address_line1'),
            address_line2=company_data.get('address_line2'),
            city=company_data.get('city'),
            state=company_data.get('state'),
            postal_code=company_data.get('postal_code'),
            country=company_data.get('country'),
            default_currency=company_data.get('default_currency', 'USD'),
            default_language=company_data.get('default_language', 'en-US'),
            timezone=company_data.get('timezone', 'UTC'),
            fiscal_year_start=company_data.get('fiscal_year_start', '01-01'),
            subscription_tier=company_data.get('subscription_tier', SubscriptionTier.BASIC),
            status=CompanyStatus.TRIAL,
            trial_ends_at=datetime.utcnow() + timedelta(days=30),
            database_schema=f"company_{company_code.lower()}",
            enabled_modules=self._get_default_modules(),
            numbering_formats=self._get_default_numbering_formats(),
            created_by=admin_user_id,
            updated_by=admin_user_id
        )
        
        self.db.add(company)
        self.db.flush()
        
        self.add_user_to_company(company.id, admin_user_id, role="admin", is_admin=True)
        self._create_default_settings(company.id)
        
        self.db.commit()
        self.db.refresh(company)
        
        return company
    
    def update_company(self, company_id: UUID, company_data: Dict[str, Any], updated_by: UUID) -> Company:
        company = self.get_company(company_id)
        if not company:
            raise ValueError(f"Company {company_id} not found")
        
        for field, value in company_data.items():
            if hasattr(company, field) and value is not None:
                setattr(company, field, value)
        
        company.updated_by = updated_by
        company.updated_at = datetime.utcnow()
        
        self.db.commit()
        self.db.refresh(company)
        
        return company
    
    def add_user_to_company(
        self, 
        company_id: UUID, 
        user_id: UUID, 
        role: str = "user", 
        is_admin: bool = False,
        permissions: Optional[Dict] = None
    ) -> CompanyUser:
        """Add User To Company."""
        """Add a user to a company."""
        existing = self.db.query(CompanyUser).filter(
            and_(
                CompanyUser.company_id == company_id,
                CompanyUser.user_id == user_id
            )
        ).first()
        
        if existing:
            raise ValueError("User is already associated with this company")
        
        company_user = CompanyUser(
            company_id=company_id,
            user_id=user_id,
            role=role,
            is_admin=is_admin,
            permissions=permissions,
            created_by=user_id,
            updated_by=user_id
        )
        
        self.db.add(company_user)
        self.db.commit()
        self.db.refresh(company_user)
        
        return company_user
    
    def get_company(self, company_id: UUID) -> Optional[Company]:
        return self.db.query(Company).filter(Company.id == company_id).first()
    
    def get_company_by_code(self, company_code: str) -> Optional[Company]:
        return self.db.query(Company).filter(Company.company_code == company_code).first()
    
    def list_companies(self, status: Optional[str] = None, limit: int = 100) -> List[Company]:
        query = self.db.query(Company)
        
        if status:
            query = query.filter(Company.status == status)
        
        return query.order_by(desc(Company.created_at)).limit(limit).all()
    
    def get_company_users(self, company_id: UUID, active_only: bool = True) -> List[CompanyUser]:
        query = self.db.query(CompanyUser).filter(CompanyUser.company_id == company_id)
        
        if active_only:
            query = query.filter(CompanyUser.is_active == True)
        
        return query.all()
    
    def get_user_companies(self, user_id: UUID, active_only: bool = True) -> List[Company]:
        query = self.db.query(Company).join(CompanyUser).filter(
            CompanyUser.user_id == user_id
        )
        
        if active_only:
            query = query.filter(
                and_(
                    Company.status == CompanyStatus.ACTIVE,
                    CompanyUser.is_active == True
                )
            )
        
        return query.all()
    
    def _generate_company_code(self, company_name: str) -> str:
        base_code = ''.join(c.upper() for c in company_name if c.isalnum())[:6]
        suffix = secrets.token_hex(2).upper()
        company_code = f"{base_code}{suffix}"
        
        while self.get_company_by_code(company_code):
            suffix = secrets.token_hex(2).upper()
            company_code = f"{base_code}{suffix}"
        
        return company_code
    
    def _get_default_modules(self) -> Dict[str, bool]:
        return {
            "general_ledger": True,
            "accounts_payable": True,
            "accounts_receivable": True,
            "payroll": False,
            "inventory": False,
            "fixed_assets": False,
            "budgeting": False,
            "reporting": True,
            "analytics": False
        }
    
    def _get_default_numbering_formats(self) -> Dict[str, str]:
        return {
            "invoice": "INV-{YYYY}-{####}",
            "purchase_order": "PO-{YYYY}-{####}",
            "journal_entry": "JE-{YYYY}-{####}",
            "payment": "PAY-{YYYY}-{####}",
            "receipt": "REC-{YYYY}-{####}",
            "quote": "QUO-{YYYY}-{####}"
        }
    
    def _create_default_settings(self, company_id: UUID):
        settings = CompanySettings(
            company_id=company_id,
            chart_of_accounts_template="standard",
            approval_workflows={
                "purchase_orders": {"enabled": False, "amount_threshold": 1000},
                "invoices": {"enabled": False, "amount_threshold": 5000},
                "journal_entries": {"enabled": False}
            },
            notification_settings={
                "email_notifications": True,
                "invoice_reminders": True,
                "payment_confirmations": True
            },
            security_settings={
                "require_2fa": False,
                "session_timeout": 30,
                "password_expiry_days": 90
            }
        )
        
        self.db.add(settings)